### ISSUE TITLE
callbacks for validations data

### ISSUE THREAD

**Issue by eyaler** [keras-team/keras#4185]
callbacks are not called when iterating over validation batches in fit methods or in evaluate methods. an example use case is resetting a stateful rnn state after some batches.

**Comment by carlthome** [comment]
If you use `fit_generator` you can reset the states in the generators, including for validation data, instead.

**Comment by eyaler** [comment]
but the generator can be asynchronous with the loop if you use queuing. also, i am not sure how to workout the scopes: would the generator modify the model as a global variable?

**Comment by carlthome** [comment]
Hmm, that's true. Good catch. Perhaps introduce new callbacks for validation in a pr?

**Comment by juliacamps** [comment]
Excuse me since I'm going a bit out of topic. But, it's more a question that an issue itself. And I think that you two know the answer.

Regarding what you pointed out of the non-synchrony between the generator and the fit loop, would it imply that managing state, by the data structure awareness (knowing which file is being used to feed the fit() at a certain moment) in stateful-RNNs is only possible by implementing the training loop (e.g. for all batches in data calling train_on_batch)? I was looking at callbacks, but ... I just realised that it might be not possible at the moment to do it by callbacks either.

Just to add some context to the question, the target problem I'm thinking about is training a temporal model (such as RNN), on some data, which are sequences of measurements from patients, the data from each patient was gathered sequentially in time, but it's to be treated as a single batch, though stateful concept should come in, but it should go away once the next patients data starts being fed to the model. I was struggling on how to design the code for achieving it, I initially thought about managing it from the generator, ... but even being a "dirty" solution, in practice, it might even not work due to the synchrony feature of generators. Is there a "nice" way to handle this state control?

Thank you in advance, and sorry for going out of topic.

**Comment by carlthome** [comment]
I'd probably do something like this

````python
sequence_length = ...  # The full length of the sequences before splitting them into RNN timesteps.

def reset_states(batch, logs):
    if batch % sequence_length == 0:
        model.reset_states()

model.fit(..., callbacks=[LambdaCallback(on_batch_end=reset_states)])
```

but in practice it doesn't seem super important to clear the cell states exactly at the end of the sequences. What seems somewhat important though is that the cell states are cleared occasionally so the counts don't increment/decrement too much.

**Comment by juliacamps** [comment]
Great, thank you!

**Comment by bstriner** [comment]
On a somewhat related note, will be starting a PR for adding callbacks to the evaluate loop. Any work on this issue so far?

https://github.com/fchollet/keras/issues/4880

**Comment by eyaler** [comment]
i believe this is still relevant

**Comment by eyaler** [comment]
Dont close bot

**Comment by eyaler** [comment]
not stale

**Comment by bzamecnik** [comment]
This is still relevant. Eg. a callback may want to measure elapsed time per batch or throughput of samples/sec. So far it works only in training mode, not validation one.

**Comment by velocirabbit** [comment]
> If you use fit_generator you can reset the states in the generators, including for validation data, instead.

Is this still possible? I just tried, and I'm getting a TF error about requiring tensors to be from the same graphs.

**Comment by timakro** [comment]
@velocirabbit Having the same issue.

**Comment by timakro** [comment]
Just found a workaround https://github.com/keras-team/keras/issues/3643#issuecomment-381971244.
